src/analyses/mallocWrapperAnalysis.ml

```diff
@@ -54,40 +54,49 @@ struct
   let threadenter ctx lval f args = [D.top ()]
   let threadspawn ctx lval f args fctx = ctx.local
   let exitstate  v = D.top ()
+  let heap_hash = ref (Hashtbl.create 113)
 
-  let heap_hash = Hashtbl.create 113
-  (* This hashtable is not marshaled, which is a problem.
-     But since Goblintutil.create_var is deterministic w.r.t the name containing location,
-     then it gets identically reconstructed when unmarshaling and verifying. *)
-
-  let get_heap_var sideg loc =
+  let get_heap_var sideg node =
     (* Use existing varinfo instead of allocating a duplicate,
        which would be equal by determinism of create_var though. *)
     (* TODO: is this poor man's hashconsing? *)
-    try Hashtbl.find heap_hash loc
+    try Hashtbl.find !heap_hash node
     with Not_found ->
-      let name = "(alloc@" ^ CilType.Location.show loc ^ ")" in
+      let name = match node with
+        | Node.Statement s -> "(alloc@" ^ "sid" ^ ":" ^ string_of_int s.sid ^ ")"
+        | Function f -> "(alloc@" ^ "vid" ^ ":" ^ string_of_int f.svar.vid ^ ")"
+        | _ -> raise (Failure "A function entry node can never be the node after a malloc") in
       let newvar = Goblintutil.create_var (makeGlobalVar name voidType) in
-      Hashtbl.add heap_hash loc newvar;
+      Hashtbl.add !heap_hash node newvar;
       sideg newvar true;
       newvar
 
   let query (ctx: (D.t, G.t, C.t) ctx) (type a) (q: a Q.t): a Queries.result =
     match q with
     | Q.HeapVar ->
-      let loc = match ctx.local with
-      | `Lifted vinfo -> vinfo
-      | _ -> Node.location ctx.node in
-      `Lifted (get_heap_var ctx.sideg loc)
+      let node = match ctx.local with
+        | `Lifted vinfo -> vinfo
+        | _ -> ctx.node in
+      `Lifted (get_heap_var ctx.sideg node)
     | Q.IsHeapVar v ->
       ctx.global v
     | Q.IsMultiple v ->
       ctx.global v
     | _ -> Queries.Result.top q
 
-    let init () =
-      List.iter (fun wrapper -> Hashtbl.replace wrappers wrapper ()) (get_string_list "exp.malloc.wrappers");
-      Hashtbl.clear heap_hash
+  let init () =
+    List.iter (fun wrapper -> Hashtbl.replace wrappers wrapper ()) (get_string_list "exp.malloc.wrappers");
+    Hashtbl.clear !heap_hash;
+    let incremental_mode = get_string "exp.incremental.mode" in
+    if incremental_mode <> "off" then (
+      match Serialize.load_heap_vars () with
+      | Some h -> heap_hash := h
+      | None -> ()
+    )
+
+  let finalize () =
+    let incremental_mode = get_string "exp.incremental.mode" in
+    if incremental_mode <> "off" then Serialize.save_heap_vars !heap_hash
```

Comment:
PR #354 adds general support for this sort of thing (instead of spreading incremental logic all over the place) and also fixes the marshaling issue here, so that one should probably be merged first.